File: 프로그래머스/1/172928. 공원 산책/공원 산책.java
import java.util.*;

class Solution {
    public static int[] solution(String[] park, String[] routes) {
        int[] answer = {0, 0};

        //시작위치 찾기
        for (int i = 0; i < park.length; i++) {
            for (int j = 0; j < park[i].length(); j++) {
                if (park[i].charAt(j) == 'S') {
                    answer[0] = i;
                    answer[1] = j;
                }
            }
        }

        Map<String, int[]> direction = new HashMap<>();
        direction.put("E", new int[]{0, 1});
        direction.put("W", new int[]{0, -1});
        direction.put("S", new int[]{1, 0});
        direction.put("N", new int[]{-1, 0});

        for (String str : routes) {
            String[] parts = str.split(" ");
            String dir = parts[0];
            int count = Integer.parseInt(parts[1]);

            int[] move = direction.get(dir);
            int nextX = answer[0];
            int nextY = answer[1];

            for (int i = 0; i < count; i++) {
                nextX += move[0];
                nextY += move[1];

                if (nextX < 0 || nextY < 0 || nextX >= park.length || nextY >= park[0].length() || park[nextX].charAt(nextY) == 'X') {
                    nextX = answer[0];
                    nextY = answer[1];
                    break;
                }
            }
            answer[0] = nextX;
            answer[1] = nextY;
        }
        return answer;
    }
}

File: 프로그래머스/1/172928. 공원 산책/공원 산책.kt
class Solution {
    fun solution(park: Array<String>, routes: Array<String>): IntArray {
        var answer: IntArray = intArrayOf(0, 0)

        // 시작위치
        park.forEachIndexed { i, s ->
            s.forEachIndexed { j, _ ->
                if (park[i][j] == 'S') {
                    answer = intArrayOf(i, j)
                }
            }
        }

        val direction = hashMapOf<String, IntArray>(
            "E" to intArrayOf(0, 1),
            "W" to intArrayOf(0, -1),
            "S" to intArrayOf(1, 0),
            "N" to intArrayOf(-1, 0)
        )

        routes.forEach { str ->
            //방향, 이동횟수 저장
            val way = direction[str.split(" ")[0]] //E
            val cnt = str.split(" ")[1].toInt() //2
            var x = answer[0]
            var y = answer[1]

            for (j in 1..cnt) {
                x += way?.get(0) ?: 0
                y += way?.get(1) ?: 0

                if (x < 0 || y < 0 || x >= park.size || y >= park[0].length || park[x][y] == 'X') {
                    x = answer[0]
                    y = answer[1]
                    break
                }
            }
            answer[0] = x
            answer[1] = y
        }
        return answer
    }
}